src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -370,61 +449,72 @@ public TStructuralType CopyChangedValues(TStructuralType original)
+        /// </summary>
+        /// <param name="original">The entity to be updated.</param>
+        public void CopyUnchangedValues(TStructuralType original)
+        {
+            if (original == null)
+            {
+                throw Error.ArgumentNull(nameof(original));
+            }
 
-                    _allProperties[nestedResourceName].SetValue(original, instance);
-                }
-                else
-                {
-                    // Recursively patch the subtree.                    
-                    Contract.Assert(TypedDelta.IsDeltaOfT(((object)deltaNestedResource).GetType()), nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+            if (!_structuredType.IsInstanceOfType(original))
+            {
+                throw Error.Argument(nameof(original), SRResources.DeltaTypeMismatch, _structuredType, original.GetType());
+            }
 
-                    Type newType = deltaNestedResource.StructuredType;
-                    Type originalType = originalNestedResource.GetType();
+            IEnumerable<PropertyAccessor<TStructuralType>> propertiesToCopy = GetUnchangedPropertyNames().Select(s => _allProperties[s]);
+            foreach (PropertyAccessor<TStructuralType> propertyToCopy in propertiesToCopy)
+            {
+                propertyToCopy.Copy(_instance, original);
+            }
 
-                    if (deltaNestedResource.IsComplexType && newType != originalType)
-                    {
-                        originalNestedResource = ReAssignComplexDerivedType(originalNestedResource, newType, originalType, deltaNestedResource.ExpectedClrType);
-                        _structuredType.GetProperty(nestedResourceName).SetValue(original, (object)originalNestedResource);
-                    }
+            CopyUnchangedDynamicValues(original);
+        }
 
-                    deltaNestedResource.CopyChangedValues(originalNestedResource);
-                }
-            }
+        /// <summary>
+        /// Overwrites the <paramref name="original"/> entity with the changes tracked by this delta.
+        /// </summary>
+        /// <remarks>The semantics of this operation are equivalent to a HTTP PATCH operation, hence the name.</remarks>
+        /// <param name="original">The entity to be updated.</param>
+        public void Patch(TStructuralType original)
+        {
+            CopyChangedValues(original);
+        }
 
-            return original;
+        /// <summary>
+        /// Overwrites the <paramref name="original"/> entity with the values stored in this delta.
+        /// </summary>
+        /// <remarks>The semantics of this operation are equivalent to a HTTP PUT operation, hence the name.</remarks>
+        /// <param name="original">The entity to be updated.</param>
+        public void Put(TStructuralType original)
+        {
+            CopyChangedValues(original);
+            CopyUnchangedValues(original);
```

Comment:
Wouldn't this call undo some of what is done by `CopyChangedValues` above?

Reply:
Should we be calling the two methods in reverse order perhaps?

Reply:
I don't see how it will undo the changes. The first method only deals with changed properties. The second method deals with unchanged properties. From the delta we can get the changedProperties and we can use that to get the unchanged properties. Am not sure how the 2 operations will affect each other.